server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/endpoint/EstimateSampleSizes.java

```diff
@@ -1,73 +1,258 @@
 
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public EndpointType endpointType() {
-        return EndpointType.GET;
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public EndpointType endpointType() {
+    return EndpointType.GET;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public String endpointName() {
+    return "/estimate-sample-sizes";
+  }
+
+  /**
+   * @return STATE authorization is necessary for this endpoint.
+   */
+  public AuthorizationType requiredAuthorization() {
+    return AuthorizationType.STATE;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected ASMEvent endpointEvent() {
+    return my_event.get();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  protected void reset() {
+    my_event.set(null);
+  }
+
+  /**
+   * Headers for the CSV file. "Ballots cast" are the number that had the contest on them;
+   * "total ballots" is the total universe size (which includes all the ballots in any county that
+   * ran the contest).
+   */
+  private static final String[] HEADERS = {
+      "County",
+      "Contest Name",
+      "Contest Type",
+      "Ballots Cast",
+      "Total ballots",
+      "Diluted Margin",
+      "Sample Size"
+  };
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public String endpointBody(final Request the_request, final Response the_response) {
+    final String prefix = "[endpointBody]";
+    LOGGER.debug(String.format("%s %s", prefix, "Started Estimate Sample sizes."));
+
+    try {
+      final String data = estimateSampleSizes();
+
+      if (data.isEmpty()) {
+        final String msg = "Could not find data for sample size estimation.";
+        LOGGER.debug(String.format("%s %s.", prefix, msg));
+        dataNotFound(the_response, msg);
+      } else {
+        the_response.header("Content-Type", "test/csv");
+        the_response.header("Content-Disposition", "attachment; filename*=UTF-8''sample_sizes.csv");
+
+        the_response.body((String.join(",", HEADERS)) + "\n" + data);
+        LOGGER.debug(String.format("%s %s.", prefix, "Completed sample size estimation"));
+        okString(the_response);
+      }
+    } catch (IllegalStateException e) {
+      // This occurs if the risk limit has not been set in the state dashboard.
+      dataNotFound(the_response, e.getMessage());
     }
+    return my_endpoint_result.get();
+  }
 
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public String endpointName() {
-        return "/estimate-sample-sizes";
+  /**
+   * Compute sample sizes for all contests for which CountyContestResults exist in the database.
+   * @return A list of string arrays containing rows with the following data: county name,
+   * contest name, contest type, single or multi-jurisdictional, ballots cast, diluted margin,
+   * and estimated sample size.
+   */
+  public String estimateSampleSizes() {
+    final String prefix = "[estimateSampleSizes]";
+    final List<EstimateData> dataRows = new ArrayList<>();
```

Comment:
Why not just declare riskLimit later when it is assigned  on line 151?

ie. final BigDecimal riskLimit = ....